pkg/reconciler/subscription/subscription.go

```diff
@@ -267,30 +261,38 @@ func (r *Reconciler) resolveReply(ctx context.Context, subscription *v1.Subscrip
 	return nil
 }
 
-func (r *Reconciler) resolveDeadLetterSink(ctx context.Context, deadLetterSink *duckv1.Destination, subscription *v1.Subscription) pkgreconciler.Event {
-	// Resolve DeadLetterSink.
-	if deadLetterSink != nil {
-		// Populate the namespace for the dead letter sink since it is in the namespace
-		if deadLetterSink.Ref != nil {
-			deadLetterSink.Ref.Namespace = subscription.Namespace
-		}
-
-		deadLetterSink, err := r.destinationResolver.URIFromDestinationV1(ctx, *deadLetterSink, subscription)
+func (r *Reconciler) resolveDeadLetterSink(ctx context.Context, s *v1.Subscription, c *eventingduckv1.Channelable) pkgreconciler.Event {
+	// resolve the Subscription's dls first, fall back to the Channels's
+	if s.Spec.Delivery != nil && s.Spec.Delivery.DeadLetterSink != nil {
+		deadLetterSinkURI, err := r.destinationResolver.URIFromDestinationV1(ctx, *s.Spec.Delivery.DeadLetterSink, s)
 		if err != nil {
+			s.Status.PhysicalSubscription.DeadLetterSinkURI = nil
 			logging.FromContext(ctx).Warnw("Failed to resolve spec.delivery.deadLetterSink",
 				zap.Error(err),
-				zap.Any("delivery.deadLetterSink", subscription.Spec.Delivery.DeadLetterSink))
-			subscription.Status.MarkReferencesNotResolved(deadLetterSinkResolveFailed, "Failed to resolve spec.delivery.deadLetterSink: %v", err)
+				zap.Any("delivery.deadLetterSink", s.Spec.Delivery.DeadLetterSink))
+			s.Status.MarkReferencesNotResolved(deadLetterSinkResolveFailed, "Failed to resolve spec.delivery.deadLetterSink: %v", err)
 			return pkgreconciler.NewEvent(corev1.EventTypeWarning, deadLetterSinkResolveFailed, "Failed to resolve spec.delivery.deadLetterSink: %w", err)
 		}
-		// If there is a change in resolved URI, log it.
-		if subscription.Status.PhysicalSubscription.DeadLetterSinkURI == nil || subscription.Status.PhysicalSubscription.DeadLetterSinkURI.String() != deadLetterSink.String() {
-			logging.FromContext(ctx).Debugw("Resolved deadLetterSink", zap.String("deadLetterSinkURI", deadLetterSink.String()))
-			subscription.Status.PhysicalSubscription.DeadLetterSinkURI = deadLetterSink
+
+		logging.FromContext(ctx).Debugw("Resolved deadLetterSink", zap.String("deadLetterSinkURI", deadLetterSinkURI.String()))
+		s.Status.PhysicalSubscription.DeadLetterSinkURI = deadLetterSinkURI
+		// In case there is no DLS defined in the Subscription Spec, fallback to Channel's
+	} else if c.Spec.Delivery != nil && c.Spec.Delivery.DeadLetterSink != nil {
+		if c.Status.DeadLetterSinkURI != nil {
```

Comment:
nit: there seems to be 3 blocks here that could be processed and returned individually. I think creating those 3 independent blocks and returning soon would be a bit more idiomatic and clear.

```
if subscription-has-dls {
    do-stuff
    return
}

if channel-has-dls {
    do-stuff
    return
}

// no dls at all
do-stuff
return
```